### PR TITLE
Switch Nixpkgs to using more NetBSD stuff for MUSL

### PR DESCRIPTION
**Pull request by matthewbauer**
###### Motivation for this change

This gets rid of some of the Musl specific stuff that was being pulled indirectly from NetBSD sources. Libbsd is also replaced with NetBSD compat. More testing is still needed.

This gets our BSD derivations directly from the upstream NetBSD CVS tree. Previously we were relying on "Linux"ized derivations of BSD sources (slightly modified versions of the original BSD sources). There are basically two things this changes:

- Replace libbsd with NetBSD compat. The Freedesktop project maintains libbsd but they only support it on Linux platforms. libbsd works great for Linux but unfortunately lacks feature detection & requires heavy patching for macOS or really anything else that's not GNU/Linux. NetBSD compat does proper feature detection so should work almost anywhere. (https://github.com/NetBSD/src/tree/trunk/tools/compat)
- Replace musl-* stuff with direct NetBSD derivation. These are all derived from NetBSD sources originally but we were relying on Alpine Linux's version of them. Moving to upstream should be more reliable.

### CODE REVIEW

**Comment by matthewbauer** [comment]
@GrahamcOfBorg build mg

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -21462,6 +21457,10 @@ with pkgs;
                       mount wall hostname more sysctl getconf
                       getent;
 
+  fts = if hostPlatform.isMusl then netbsd.fts else null;
```
Comment (Ericson2314) [inline pkgs/top-level/all-packages.nix]: Maybe there is some `meta.available` thing you can do instead of `null`?

Comment (matthewbauer) [reply pkgs/top-level/all-packages.nix]: It should be provided by most libcs except for Musl. It will build everywhere, just runs on Musl.

**Comment by edolstra** [comment]
I don't understand what this means. Why would we want to use "more NetBSD stuff"? If the goal is to move away from GNU, :-1:.

**Comment by matthewbauer** [comment]
> I don't understand what this means. Why would we want to use "more NetBSD stuff"

I should have been more clear. This gets our BSD derivations directly from the upstream NetBSD CVS tree. Previously we were relying on "Linux"ized derivations of BSD sources. There are basically two things this changes:

- Replace libbsd with NetBSD compat. The Freedesktop project maintains libbsd but they only support it on Linux platforms. libbsd works great for Linux but unfortunately lacks feature detection & requires heavy patching for macOS or really anything else that's not GNU/Linux. NetBSD compat does proper feature detection so should work almost anywhere. (https://github.com/NetBSD/src/tree/trunk/tools/compat)
- Replace musl-* stuff with direct NetBSD derivation. These are all derived from NetBSD sources originally but we were relying on Alpine Linux's version of them. Moving to upstream should be more reliable.

**Comment by Ericson2314** [comment]
@edolstra yeah even more pithily put, this is *not* moving away from GNU, but just using upstream version of the non-GNU stuff we already use.

**Comment by vcunat** [comment]
This broke samba build, now discussed on #nixos-dev with @7c6f434c.

**Comment by orivej** [comment]
This have broken all packages that depend on libbsd, and the reason is obvious if you compare the `include` files that they install:
```
# freedesktop libbsd      # netbsd compat
└── bsd                   ├── cdbr.h
    ├── bitstring.h       ├── cdbw.h
    ├── bsd.h             ├── compat_defs.h
    ├── err.h             ├── db.h
    ├── getopt.h          ├── machine
    ├── inttypes.h        │   ├── bswap.h
    ├── libutil.h         │   └── endian.h
    ├── md5.h             ├── md5.h
    ├── netinet           ├── nbtool_config.h
    │   └── ip_icmp.h     ├── netconfig.h
    ├── nlist.h           ├── rmd160.h
    ├── readpassphrase.h  ├── rpc
    ├── stdio.h           │   └── types.h
    ├── stdlib.h          ├── sha1.h
    ├── string.h          ├── sha2.h
    ├── stringlist.h      ├── sys
    ├── sys               │   ├── bswap.h
    │   ├── bitstring.h   │   ├── endian.h
    │   ├── cdefs.h       │   ├── featuretest.h
    │   ├── endian.h      │   ├── queue.h
    │   ├── poll.h        │   └── tree.h
    │   ├── queue.h       ├── tzfile.h
    │   ├── time.h        ├── util.h
    │   └── tree.h        ├── utmpx.h
    ├── timeconv.h        └── vis.h
    ├── unistd.h
    ├── vis.h
    └── wchar.h
```
[acmeclient](https://hydra.nixos.org/build/75436483):
```
config.h:21:12: fatal error: bsd/stdlib.h: No such file or directory
```
[hfsprogs](https://hydra.nixos.org/build/75438460):
```
/build/diskdev_cmds-332.25/include/missing.h:9:10: fatal error: bsd/string.h: No such file or directory
```
[radiotray-ng](https://hydra.nixos.org/build/75427147):
```
-- Checking for one of the modules 'libbsd'
CMake Error at /nix/store/5adacq979k95wi13q6fzhdddgxbqnpz0-cmake-3.11.2/share/cmake-3.11/Modules/FindPkgConfig.cmake:640 (message):
  None of the required 'libbsd' found
```
where `radiotray-ng` `CMakeLists.txt` contains
```
if (NOT APPLE)
    pkg_search_module(LIBBSD  REQUIRED libbsd)
endif(NOT APPLE)
```
Clearly nbcompat can not replace libbsd.

**Comment by matthewbauer** [comment]
Yes if someone wants to undo the ```libbsd = netbsd.compat``` they are certainly welcome to. A few do build but mostly just ones using basic symbols like strlcpy. I still want to minimize libbsd usage in nixpkgs but it is obviously not a drop in replacement right now.

**Comment by orivej** [comment]
OK, I'll undo `libbsd = netbsd.compat`.

I have noticed that having a direct or an indirect dependency on `netbsd.compat` adds `-include nbtool_config.h` to `NIX_CFLAGS_COMPILE`, which prepends a lot of stuff to every compiled file, such as:
```
#define HAVE_ASPRINTF 1
#define HAVE_FEATURES_H 1
#define STDC_HEADERS 1

#include "compat_defs.h"

# ← from compat_defs.h

#if defined(__linux__) && HAVE_FEATURES_H
#include <features.h>
#undef _POSIX_SOURCE
#undef _POSIX_C_SOURCE
#define __USE_ISOC99 1
#endif

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#undef btodb
#define btodb(x) ((x) << 9)
#undef setbit
#define	setbit(a,i)	((a)[(i)/NBBY] |= 1<<((i)%NBBY))
```
and e.g. this file compiles and runs:
```
int main() {
    puts("hello");
    return btodb(10);
}
```

Is this `-include` necessary? If it is, can it be limited to direct dependencies? It has high potential to introduce subtle breakages.

**Comment by matthewbauer** [comment]
It may not be necessary any more but it definitely was at one point. The NetBSD derivations are built on top of this so they need those defines to avoid compatibility issues. This is supposed to get you to a BSD-compatible system. I can look into either disabling the default or making it optional though.

**Comment by orivej** [comment]
If only stuff under `netbsd` will depend on `netbsd.compat`, then it seems OK to leave this as is.

**Comment by orivej** [comment]
Sorry, I was wrong about indirect dependencies in:
> I have noticed that having a direct or an indirect dependency on `netbsd.compat` adds `-include nbtool_config.h` to `NIX_CFLAGS_COMPILE`
> Is this `-include` necessary? If it is, can it be limited to direct dependencies?
I have tested this with a package that indirectly depended on `netbsd.fts` which propagates `netbsd.compat`.

However, a package with `-include nbtool_config.h` in its setupHook should not be propagated. The packages that currently propagate it should have the same setupHook instead. I'm going to propose this in a PR.

**Pull request merged**